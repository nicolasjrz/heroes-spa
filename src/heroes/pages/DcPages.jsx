import React from "react";
import { HeroList } from "../components";
export const DcPages = () => {
  return (
    <>
      <h1>DC comics</h1>
      <hr />

      <HeroList publisher="DC Comics" />
    </>
  );
};
