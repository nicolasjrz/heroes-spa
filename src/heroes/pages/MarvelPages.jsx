import React from "react";
import { HeroList } from "../components";

export const MarvelPages = () => {
  return (
    <>
      <h1>Marvel comics</h1>
      <hr />

      <HeroList publisher="Marvel Comics" />
    </>
  );
};
