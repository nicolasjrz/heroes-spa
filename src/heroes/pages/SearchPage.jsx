import React from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import queryString from "query-string";
import { getSearchHero } from "../helpers/getSearchHero";
import { HeroCard } from "../components";
export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);
  const hero = getSearchHero(q);
  const { searchText, onInputChange } = useForm({ searchText: q });

  const showSearch = q.length === 0;
  const showError = q.length > 0 && hero.length === 0;

  const onSearchSubmit = (e) => {
    e.preventDefault();

    navigate(`?q=${searchText}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              name="searchText"
              placeholder="search a hero"
              className="form-control"
              autoComplete="off"
              onChange={onInputChange}
              value={searchText}
            />
            <button className="btn btn-primary mt-3">Search</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Result</h4>
          <hr />

          {/* {q === "" ? (
            <div className="alert alert-primary">search a hero</div>
          ) : (
            hero.length === 0 && (
              <div className="alert alert-danger">no result {q}</div>
            )
          )}*/}

          <div
            className="alert alert-primary animate__animated animate__flipInX"
            style={{ display: showSearch ? "" : "none" }}
          >
            search a hero
          </div>
          <div
            className="alert alert-danger animate__animated animate__flipInX"
            style={{ display: showError ? "" : "none" }}
          >
            no result {q}
          </div>
          <div className="">
            {hero.map((hero) => (
              <HeroCard key={hero.id} heroe={hero} />
            ))}
          </div>
        </div>
      </div>
    </>
  );
};
