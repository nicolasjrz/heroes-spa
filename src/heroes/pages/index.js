export * from "./DcPages";
export * from "./MarvelPages";
export * from "./HeroPage";
export * from "./SearchPage";
