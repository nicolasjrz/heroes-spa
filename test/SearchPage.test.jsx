import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../src/heroes/pages";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("pruebas en Search Page", () => {
  beforeEach(() => jest.clearAllMocks());
  test("debe mostrarse correctamente con los valores por defecto", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("debe mostrarse a batman y el input con el valor del quyeryString", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    expect(input.value).toBe("batman");

    const img = screen.getByRole("img");
    expect(img.src).toContain("/assets/heroes/dc-batman.jpg");

    const alert = screen.getByLabelText("alert-hero");

    expect(alert.style.display).toBe("none");
  });

  test("debe mostrarse un mensaje si no se encuentra el heroe", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batmanasdasd"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const alert = screen.getByLabelText("alert-hero");

    expect(alert.style.display).toBe("");
  });

  test("dene de llamar el navigate a la pantalla nueva", () => {
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "Spider Man" } });

    const onsubmit = screen.getByRole("button");
    fireEvent.click(onsubmit);

    expect(mockedUseNavigate).toHaveBeenCalledWith("?q=Spider Man");
  });
});
