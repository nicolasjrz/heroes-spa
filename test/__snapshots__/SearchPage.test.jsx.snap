// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en Search Page debe mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__flipInX"
      >
        search a hero
      </div>
      <div
        aria-label="alert-hero"
        class="alert alert-danger animate__animated animate__flipInX"
        style="display: none;"
      >
        no result 
      </div>
      <div
        class=""
      />
    </div>
  </div>
</div>
`;
